### PR TITLE
Add versioned task for resolver task

### PR DESCRIPTION
**Pull request by savitaashture**
# Changes

* Added changes to show versioned tasks for resolver task
   *  Fixes: https://issues.redhat.com/browse/SRVKP-5595
* Refactored to avoid code duplication

Result:
```
$ kubectl get task -n openshift-pipelines
NAME                      AGE
buildah                   5m15s
buildah-1-16-0            5m18s
git-cli                   5m15s
git-cli-1-16-0            5m18s
```

new installerset to handle versioned resolver task
```
$ kubectl get tektoninstallersets
NAME                                             READY   REASON
....
addon-versioned-resolvertasks-devel-hx55x        True   
```

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Run `make test lint` before submitting a PR
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

File: pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go [hunk 1]
```diff
@@ -28,14 +28,11 @@ import (
...
-)
-
-// VersionedClusterTaskSet this is an exception case where we create one installer set for one minor version
-// not for patch version and we don't remove older installer sets on upgrade, hence keeping it different
+// VersionedTaskSet this is an exception case where we create one installer set for one minor version
+// not for patch version, and we don't remove older installer sets on upgrade, hence keeping it different
 // from custom set otherwise code becomes unnecessarily complex to handle this case
-func (i *InstallerSetClient) VersionedClusterTaskSet(ctx context.Context, comp v1alpha1.TektonComponent, manifest *mf.Manifest, filterAndTransform FilterAndTransform) error {
+func (i *InstallerSetClient) VersionedTaskSet(ctx context.Context, comp v1alpha1.TektonComponent, manifest *mf.Manifest,
+	filterAndTransform FilterAndTransform, isType, isName string) error {
```
Comment (jkandasa) [inline pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go:35]: nit: generally I do not prefer to use `is` as prefix for non `bool` variables
May be,
`isType` => `insType`
`isName` => `insName`

File: pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go [hunk 2]
```diff
@@ -48,29 +45,29 @@ func (i *InstallerSetClient) VersionedClusterTaskSet(ctx context.Context, comp v
...
 
 	if len(is.Items) == 0 {
-		vctSet, err := i.makeInstallerSet(ctx, comp, manifestUpdated, "addon-versioned-clustertasks", setType, nil)
+		vctSet, err := i.makeInstallerSet(ctx, comp, manifestUpdated, isName, setType, nil)
 		if err != nil {
 			return err
 		}
 		vctSet.Labels[v1alpha1.ReleaseMinorVersionKey] = getPatchVersionTrimmed(i.releaseVersion)
-		vctSet.GenerateName = "addon-versioned-clustertasks" + "-" + getPatchVersionTrimmed(i.releaseVersion) + "-"
+		vctSet.GenerateName = isName + "-" + getPatchVersionTrimmed(i.releaseVersion) + "-"
```
Comment (jkandasa) [inline pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go:70]: nit: I prefer to use `fmt.Sprintf()`
```
vctSet.GenerateName = fmt.Sprintf("%s-%s-", isName, getPatchVersionTrimmed(i.releaseVersion))
```

File: pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolver-name-expected.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# auto generated by hack/openshift/fetch-tektoncd-catalog-tasks.sh
+# DO NOT EDIT: use the script instead
+# source: https://raw.githubusercontent.com/openshift/tektoncd-catalog/release-v0.10/buildah/buildah.yaml
+#
```
Comment (jkandasa) [inline pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolver-name-expected.yaml:4]: I guess this test data is not managed by `hack/openshift/fetch-tektoncd-catalog-tasks.sh`?

File: pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolvertask-name.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# auto generated by hack/openshift/fetch-tektoncd-catalog-tasks.sh
+# DO NOT EDIT: use the script instead
+# source: https://raw.githubusercontent.com/openshift/tektoncd-catalog/release-v0.10/buildah/buildah.yaml
+#
+---
```
Comment (jkandasa) [inline pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolvertask-name.yaml:5]: I guess this test data is not managed by `hack/openshift/fetch-tektoncd-catalog-tasks.sh`?

**Review by jkandasa — COMMENTED** [review]
@savitaashture Please change the go file names from `camelCase` to `snake_case`

File: pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go (hunk 1 above)
Comment (savitaashture) [reply pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go]: Updated

File: pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go (hunk 2 above)
Comment (savitaashture) [reply pkg/reconciler/kubernetes/tektoninstallerset/client/custom_versioned_clustertask.go]: Updated

File: pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolver-name-expected.yaml (hunk 3 above)
Comment (savitaashture) [reply pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolver-name-expected.yaml]: Yes 
i copied from another example where it was mentioned
but yeah removed those comments

File: pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolvertask-name.yaml (hunk 4 above)
Comment (savitaashture) [reply pkg/reconciler/openshift/tektonaddon/testdata/test-versioned-resolvertask-name.yaml]: removed comments

**Comment by savitaashture** [comment]
> @savitaashture Please change the go file names from `camelCase` to `snake_case`

Changed all those files to snake_case from tektonaddon folders

File: pkg/reconciler/openshift/tektonaddon/cluster_task_versioned.go [hunk 5]
```diff
@@ -22,13 +22,20 @@ import (
...
-		if err := r.installerSetClient.VersionedClusterTaskSet(ctx, ta, &manifest, filterAndTransformVersionedClusterTask(r.operatorVersion)); err != nil {
+		addonImages := common.ToLowerCaseKeys(common.ImagesFromEnv(common.AddonsImagePrefix))
+		tfs := []mf.Transformer{
+			replaceKind(KindTask, KindClusterTask),
+			injectLabel(labelProviderType, providerTypeRedHat, overwrite, KindClusterTask),
+			common.TaskImages(addonImages),
+			setVersionedNames(r.operatorVersion),
+		}
+		if err := r.installerSetClient.VersionedTaskSet(ctx, ta, &manifest, filterAndTransformClusterTask(tfs),
+			"VersionedClusterTask", installerSetNameForClusterTasks); err != nil {
```
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/cluster_task_versioned.go:38]: can we use the constant here?

File: pkg/reconciler/openshift/tektonaddon/resolver_task_versioned.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	manifest := *r.resolverTaskManifest
+	if enable == "true" {
+		addonImages := common.ToLowerCaseKeys(common.ImagesFromEnv(common.AddonsImagePrefix))
+		tfs := []mf.Transformer{
+			injectLabel(labelProviderType, providerTypeRedHat, overwrite, KindTask),
+			common.TaskImages(addonImages),
+			setVersionedNames(r.operatorVersion),
+		}
+		if err := r.installerSetClient.VersionedTaskSet(ctx, ta, &manifest, filterAndTransformResolverTask(tfs),
+			"VersionedResolverTask", installerSetNameForResolverTasks); err != nil {
```
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/resolver_task_versioned.go:37]: can we use constant here?

**Review by jkandasa — APPROVED** [review]
cc @piyush-garg

**Pull request merged**